EC2 instance metadata implementation.

spasam opened the pull request:
This is based on https://github.com/facebook/osquery/pull/2853
I made some optimizations and cleanup.

When queried on non-EC2 instance, this table responds with no results
in 3 seconds. Subsequent queries are immediately returned.

On EC2 instance, each query makes 7 round trips to meta-data servers.

Original pull request had copyright claim by jfdive with BSD license.
I changed it to standard Facebook copyright.

spasam commented:
Output from non-EC2 instance. First call takes 3 seconds. Second is instantaneous:
```
osquery> select * from aws_metadata;
osquery> select * from aws_metadata;
```

Sample output from EC2 instance (with masked values):
```
osquery> select * from aws_metadata;
+---------------------+---------------+--------------+-----------+-------------------+----------------------------------------+------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------+--------------+---------------------+--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| instance_id         | instance_type | architecture | region    | availability_zone | local_hostname                         | local_ipv4 | mac               | security_groups    | iam_arn                                            | ami_id       | reservation_id      | account_id   | ssh_public_key                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
+---------------------+---------------+--------------+-----------+-------------------+----------------------------------------+------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------+--------------+---------------------+--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| i-xxxxxxxxxxxxxxxxx | r4.2xlarge    | x86_64       | us-west-2 | us-west-2a        | ip-1-0-2-6.internal.example.com        | 10.0.2.6   | 06:ab:cd:ef:00:aa | database,ssh,blah  | arn:aws:iam::000000000000:instance-profile/Role    | ami-12345677 | r-00000000000000000 | 000000000000 | ssh-rsa abcdef... keyname, |
+---------------------+---------------+--------------+-----------+-------------------+----------------------------------------+------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------+--------------+---------------------+--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
osquery>
```

spasam commented:
Here is reference documentation. There are some out-dated information: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

Thinking about it a bit more, may be this should be ec_instance_metadata as AWS is too generic!!!

obelisk commented:
ok to test

osqueryer commented:
:-1: The commit cfd5141a848b015124ff133cb9f16d040f0ddf32 (Job results: [307](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/307/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit cfd5141a848b015124ff133cb9f16d040f0ddf32 (Job results: [1455](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1455/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit cfd5141a848b015124ff133cb9f16d040f0ddf32 (Job results: [2240](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2240/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit cfd5141a848b015124ff133cb9f16d040f0ddf32 (Job results: [2241](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2241/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit cfd5141a848b015124ff133cb9f16d040f0ddf32 (Job results: [308](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/308/)) failed one or more tests (**FreeBSD**).

spasam commented:
There is some overlap between this pull request and https://github.com/facebook/osquery/pull/3507
I will fix the build issues once 3507 is addressed

theopolis commented on osquery/tables/cloud/ec2_instance_metadata.cpp:
```diff
@@ -0,0 +1,283 @@
+  SimpleEc2MetaData(const ColumnType sqlType,
+                    const std::string columnName,
+                    const std::string urlSuffix)
+      : Ec2MetaData(urlSuffix), sql_type_(sqlType), column_name_(columnName) {}
```
Do you need to `std::move` the `columnName`?

theopolis commented on osquery/tables/cloud/ec2_instance_metadata.cpp:
```diff
@@ -0,0 +1,283 @@
+   * @param http_body content of the http response body
+   * @param r The row to which the value need to be added
+   */
+  virtual void extractResult(const std::string& http_body, Row& r) const = 0;
```
Can this be `protected`?

spasam replied:
Fixed

theopolis commented on osquery/tables/cloud/ec2_instance_metadata.cpp:
```diff
@@ -0,0 +1,283 @@
+  const std::vector<std::string> key_names_;
+
+ public:
+  JSONEc2MetaData(const std::vector<std::string> columnNames,
```
For these parameters, is the intent to copy then move them into place? You may need to call `std::move` for the local ctor calls.

theopolis commented on osquery/tables/cloud/ec2_instance_metadata.cpp:
```diff
@@ -0,0 +1,283 @@
+    break;
+  }
+  case INTEGER_TYPE: {
+    if (value == "") {
```
If `value.empty()`

theopolis replied:
Maybe collapse this into a ternary

theopolis commented on osquery/tables/cloud/ec2_instance_metadata.cpp:
```diff
@@ -0,0 +1,283 @@
+    return results;
+  }
+
+  const static std::vector<Ec2MetaData*> fields(
```
Should these be smart pointers so that they are freed correctly during global deconstruction?

spasam commented:
@theopolis Thanks for your comments. Looking/fixing...

mime-frame commented:
@spasam - do not include iam/security-credentials, that will leak credentials

spasam commented:
@theopolis I think I addressed your review comments. LMK

Regarding detecting EC2 instance it looks like there is a simpler way. On EC2 instances first three characters of "/sys/hypervisor/uuid" are "ec2". I will update metadata and tags code to use this in separate pull request.

spasam commented:
For now, I added docs in configuration.md

osqueryer commented:
:-1: The commit e32fa236e5ddfab6e4305c308380681943af4b25 (Job results: [2310](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2310/)) failed one or more tests (**Windows**).
